src/app/problem.cljs

```diff
@@ -16,114 +19,180 @@
 (defn next-problem [id]
   (some #(when (> (:id %) id) %) data/problems))
 
+(defn join-human [xs]
+  (if (= 1 (count xs))
+    (first xs)
+    (str (str/join ", " (butlast xs))
+         " and "
+         (last xs))))
+#_(join-human ['a 'b 'c])
+
+(defn restricted-symbol-msg [restricted-symbols]
+  (str (if (= 1 (count restricted-symbols))
+         (str (first restricted-symbols) " is")
+         (str (join-human restricted-symbols) " are"))
+       " not allowed for this problem"))
+
 (defn check-solution [problem user-solution]
   (let [replaced (mapv #(str/replace % "__" user-solution) (:tests problem))
         results  (map eval-string replaced)
         passed   (count (filter true? results))
-        failed   (count (filter false? results))]
-    (swap! user-data assoc (:id problem) {:code   user-solution
-                                          :passed passed
-                                          :failed failed})
-    (if (or (> failed 0) (> passed 0))
-      results
-      (throw (ex-info "Evaluation error" {:stacktrace (first results)})))))
+        failed   (count (filter false? results))
+        restricted-symbols (problem-restrictions/find-restricted-symbols
+                            problem
+                            user-solution)]
+    (if restricted-symbols
+      [results (restricted-symbol-msg restricted-symbols)]
+      (do (swap! user-data assoc (:id problem) {:code   user-solution
+                                                :passed passed
+                                                :failed failed})
+          (if (or (> failed 0) (> passed 0))
+            [results nil]
+            (throw (ex-info "Evaluation error" {:stacktrace (first results)})))))))
 
 (def results-style {:width "100%"
                     :display "flex"
                     :font-family "monospace"
                     :justify-content "space-between"})
 
-(defn modal-results-section [results tests id]
-  (let [test-attempts (map vector tests results)]
-    [:div
-     [:h4 (str "Results " "#" id )]
-     [:hr ]
+(defn render-result-item [test maybe-result]
+  [:li
+   [:pre
+    [:code {:style results-style}
+     [:span {:style {:overflow-x "hidden"}} test]
+     (when maybe-result maybe-result)]]])
+
+(defn result-list [items]
+  [:ul {:style {:list-style "none" :padding 0}}
+   items])
+
+(defn test-list-section [tests]
+  [result-list
+   (doall
+    (for [[i test] (map-indexed vector tests)]
+      ^{:key i}
+      [render-result-item test]))])
+
+(defn result-info-item [color text]
+  [:span {:style {:color color
+                  :align-self "center"}}
+   text])
+
+(defn test-results-section [results tests]
+  [result-list
+   (let [test-attempts (map vector tests results)]
      (for [[i [test passed?]] (map-indexed vector test-attempts)]
        ^{:key i}
-       [:p {:style results-style}
-        [:span test]
+       [render-result-item
+        test
         (if passed?
```

Comment:
This check was bad because on error (let's say we call `([inc] 1 2 3 4 5)` the custom eval returns a string:

```
(defn eval-string [source]
  (try (sci/eval-string* context source)
       (catch :default e
         (with-out-str (error-handler source e)))))
```

Reply:
OK, let's make a separate issue for this

Reply:
I did: https://github.com/oxalorg/4ever-clojure/issues/78